File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        lst = s.split(" ")
        temp = ""
        n = len(lst)
        lst = list(reversed(lst))
        print(lst)
        for i in range(n):
            if lst[i] != " " and lst[i] != "":
                temp += lst[i]
                if i != n - 1:
                    temp += " "
            
        answer = temp
        answer = answer.strip()
        return answer

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        s = deleteEmptyStringFromStartAndBack(s);
        s = deleteOverSpace(s);
        
        String[] words = new String[s.length()];
        words = s.split(" ");
        
        List<String> answer = new ArrayList<>();
        for(int i=words.length-1; i>=0; i--){
            answer.add(words[i]);
        }
        
        s = String.join(" ", answer);
        return s;
    }
    
    private String deleteOverSpace(String s){
        s = s.replaceAll("\\s+", " ");
        return s;
    }
    
    private String deleteEmptyStringFromStartAndBack(String s){
        int idx = -1;
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) != ' '){
                idx = i;
                break;
            }
        }
        s = s.substring(idx, s.length());
        
        int ridx = s.length()-1;
        for(int i=s.length()-1; i>=0; i--){
            if(s.charAt(i) != ' '){
                ridx = i;
                break;
            }
        }
        s = s.substring(0, ridx+1);
        
        return s;
    }
}